xdsl/transforms/mpi_to_llvm.py

```diff
@@ -0,0 +1,525 @@
+                [u]i16  -> MPI_[UNSIGNED]_SHORT
+                [u]i32  -> MPI_UNSIGNED / MPI_INT
+                [u]i64  -> MPI_UNSIGNED_LONG_LONG / MPI_LONG_LONG_INT
+        """
+        if isinstance(typ, builtin.Float32Type):
+            return self.info.MPI_FLOAT
+        if isinstance(typ, builtin.Float64Type):
+            return self.info.MPI_DOUBLE
+        if isinstance(typ, IntegerType):
+            width: int = typ.width.data
+            if typ.signedness.data == Signedness.UNSIGNED:
+                # unsigned branch
+                if width == 8:
+                    return self.info.MPI_UNSIGNED_CHAR
+                if width == 16:
+                    return self.info.MPI_UNSIGNED_SHORT
+                if width == 32:
+                    return self.info.MPI_UNSIGNED
+                if width == 64:
+                    return self.info.MPI_UNSIGNED_LONG_LONG
+            else:
+                if width == 8:
+                    return self.info.MPI_CHAR
+                if width == 16:
+                    return self.info.MPI_SHORT
+                if width == 32:
+                    return self.info.MPI_INT
+                if width == 64:
+                    return self.info.MPI_LONG_LONG_INT
+            raise ValueError(
+                "MPI Datatype Conversion: Unsupported integer bitwidth: {}".
+                format(width))
+        raise ValueError(
+            "MPI Datatype Conversion: Unsupported type {}".format(typ))
+
+    def _mpi_name(self, op: mpi.MPIBaseOp) -> str:
+        """
+        Convert the name of an mpi dialect operation to the corresponding MPI function call
+        """
+        if op.name not in self.MPI_SYMBOL_NAMES:
+            raise RuntimeError(
+                "Lowering of MPI Operations failed, missing lowering for {}!".
+                format(op.name))
+        return self.MPI_SYMBOL_NAMES[op.name]
+
+    def _memref_get_llvm_ptr(
+            self, ref: SSAValue) -> tuple[list[Operation], Operation]:
+        """
+        Converts an SSA Value holding a reference to a memref to llvm.ptr
+
+        The official way as per the documentations pecifies the following
+        sequence of operations:
+
+          %0 = memref.extract_aligned_pointer_as_index %arg : memref<4x4xf32> -> index
+          %1 = arith.index_cast %0 : index to i64
+          %2 = llvm.inttoptr %1 : i64 to !llvm.ptr<f32>
+
+        https://mlir.llvm.org/docs/Dialects/MemRef/#memrefextract_aligned_pointer_as_index-mlirmemrefextractalignedpointerasindexop
+        """
+        return [
+            index := memref.ExtractAlignedPointerAsIndexOp.get(ref),
+            i64 := arith.IndexCastOp.get(index, builtin.i64),
+            ptr := llvm.IntToPtrOp.get(i64),
+        ], ptr
+
+
+class LowerMpiInit(_MPIToLLVMRewriteBase):
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: mpi.Init, rewriter: PatternRewriter, /):
+        rewriter.replace_matched_op(*self.lower(op))
+
+    def lower(self,
```

Comment:
I think this is an interesting design. The core advantage I can see over inlining this into `match_and_rewrite` is that you can now test from pytest.

Personally, I usually test rewrites in filecheck tests, because that really is what matters to me, IMO (I feed it one IR, does it properly give me the other one). So I don't really need it. Also, I think this design might not generalize to more complex rewrites that are not "expansion-based", so not of the form "replace the matched op with all these ops". I would expect that you will hit a roadblock in the lowering with this design, once you expand it enough. But maybe, that's something to think about when it actually happens :)

Reply:
Yeah, I would personally just test these things in filecheck, removing the need for this. But I don't think it's bad to have this design here, so we can keep it.